lib/Sema/CSSimplify.cpp

```diff
@@ -1230,8 +1250,85 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
   auto argumentLocator = locator.withPathElement(
       ConstraintLocator::FunctionArgument);
 
-  if (func1Params.size() != func2Params.size())
-    return getTypeMatchFailure(argumentLocator);
+  int diff = func1Params.size() - func2Params.size();
+  if (diff != 0) {
+    if (!shouldAttemptFixes())
+      return getTypeMatchFailure(argumentLocator);
+
+    auto *anchor = locator.trySimplifyToExpr();
+    if (!anchor)
+      return getTypeMatchFailure(argumentLocator);
+
+    // If there are missing arguments, let's add them
+    // using parameter as a template.
+    if (diff < 0) {
+      // If there are N parameters but a single closure argument
+      // (which might be anonymous), it's most likely used as a
+      // tuple e.g. `$0.0`.
+      Optional<TypeBase *> argumentTuple;
+      if (isa<ClosureExpr>(anchor) && isSingleParam(func1Params)) {
+        auto isParam = [](const Expr *expr) {
+          if (auto *DRE = dyn_cast<DeclRefExpr>(expr)) {
+            if (auto *decl = DRE->getDecl())
+              return isa<ParamDecl>(decl);
+          }
+          return false;
+        };
+
+        const auto &arg = func1Params.back();
+        if (auto *argTy = arg.getPlainType()->getAs<TypeVariableType>()) {
```

Comment:
Do you want to check that arg is not inout or variadic?

Reply:
That's already checked by `isSingleParam(func1Params)` in enclosing `if`.